app/services/api/v3/filter_flows.rb

```diff
@@ -0,0 +1,337 @@
+          @errors << 'Both start and end date not given'
+        end
+        @errors << 'Resize quant not given' unless @resize_quant
+        return unless @recolor_ind && @recolor_qual
+        @errors << 'Either ind or qual for recoloring'
+      end
+
+      def initialize_node_types
+        @node_types = Api::V3::NodeType.
+          select(
+            [
+              'node_types.id AS id',
+              'node_types.name AS name',
+              'context_node_types.column_position',
+              'context_node_type_properties.column_group'
+            ]
+          ).
+          joins(context_node_types: :context_node_type_property).
+          where('context_node_types.context_id' => @context.id).
+          order('context_node_types.column_position ASC')
+        @errors << 'No node types for context' unless @node_types.any?
+      end
+
+      # columns to be displayed on frontend
+      # it should be possible to support variable number of columns
+      # but it should not be possible to select more than one column from
+      # same column group
+      def initialize_active_node_types
+        @active_node_types = @node_types.select do |c|
+          @node_types_ids.include?(c['id'])
+        end
+        # TODO: do not allow to select more than one column from same group
+        @active_node_types_positions = @active_node_types.map(&:column_position)
+      end
+
+      def initialize_biome_position
+        @biome_position = @biome_id && @node_types.index do |c|
+          c['name'] == NodeTypeName::BIOME
+        end + 1
+      end
+
+      # nodes selected by the user
+      def initialize_selected_nodes
+        # filter out nodes not involved in any flows in this context
+        @selected_nodes_ids = @selected_nodes_ids.select do |node_id|
+          Api::V3::Flow.
+            select('true').
+            joins(:flow_quants).
+            where('flow_quants.quant_id' => @resize_quant.id).
+            where(context_id: @context.id). # TODO: verify this
+            where('? = ANY(flows.path)', node_id).
+            where('year >= ? AND year <= ?', @year_start, @year_end).any?
+        end
+
+        @selected_nodes = Node.where(id: @selected_nodes_ids).all
+
+        # filter out nodes which have different node types
+        @selected_nodes = @selected_nodes.select do |node|
+          @node_types_ids.include?(node.node_type_id)
+        end
+
+        @selected_nodes_by_position = @selected_nodes.group_by do |node|
+          @active_node_types.find do |nt|
+            nt.id == node.node_type_id
+          end.column_position
+        end
+      end
+
+      def initialize_other_nodes_ids
+        @other_nodes_ids = @active_node_types.map do |node_type|
+          Api::V3::Node.where(node_type_id: node_type.id, name: 'OTHER').
+            pluck(:id).first
+          # TODO: maybe we could not rely on those nodes in db?
+          # problem is the id of those nodes is referenced
```

Comment:
I meant they're referenced in the responses, as the numeric identifier is used in the responses. I guess a workaround could be to assign artificial ids to those, e.g. negative. I didn't want to include such changes here since that would change the response and make it more difficult to test.

Reply:
:+1: